strutil/intersection.go

```diff
@@ -0,0 +1,67 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// Intersection computes the intersection of a set of slices, treating each
+// slice as a set. It does not mutate any of the input slices and returns a new
+// slice. It is recursive.
+func Intersection(slices ...[]string) []string {
+	// handle trivial cases
+	switch len(slices) {
+	case 0:
+		return nil
+	case 1:
+		return slices[0]
+	case 2:
+		// actually perform the intersection
+		l1 := slices[0]
+		l2 := slices[1]
+		guessLen := len(l1)
+		if len(l1) > len(l2) {
+			guessLen = len(l2)
+		}
+		alreadyAdded := map[string]bool{}
+		result := make([]string, 0, guessLen)
+		for _, item := range l1 {
+			if ListContains(l2, item) {
+				if !alreadyAdded[item] {
```

Comment:
This could be slightly optimised by reversing the order of the two if statements since ListContains is O(n). I know performance isn't a priority here but just thought I'd suggest it since it's a small change.

I'd also merge the two statements since it'd be easier to read and less nested (but that's subjective so feel free to ignore)